designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,627 @@
+            custom: ruleLoader.loadFromDirectory("./custom-rules")
+        },
+        rules: {
+            "custom/my-rule": "error"
+        }
+    };
+
+    if (context.hasRule("some-new-rule")) {
+        myConfig.rules["some-new-rule"] = ["error"];
+    }
+
+    return myConfig;
+};
+```
+
+### Configuration Location Resolution
+
+When ESLint is executed, the following steps are taken to find the `eslint.config.js` file to use:
+
+1. If the `-c` flag is used then the specified configuration file is used. There is no further search performed.
+1. Otherwise:
+    a. Look for `eslint.config.js` in the current working directory. If found, stop searching and use that file.
+    b. If not found, search up the directory hierarchy looking for `eslint.config.js`.
+    c. If a `eslint.config.js` file is found at any point, stop searching and use that file.
+    d. If `/` is reached without finding `eslint.config.js`, then stop searching and output a "no configuration found" error.
+
+This approach will allow running ESLint from within a subdirectory of a project and get the same result as when ESLint is run from the project's root directory (the one where `eslint.config.js` is found).
+
+Some of the key differences from the way ESLint's configuration resolution works today are:
+
+1. There is no automatic search for `eslint.config.js` in the user's home directory. Users wanting this functionality can either pass a home directory file using `-c` or manually read in that file from their `eslint.config.js` file.
+1. Once a `eslint.config.js` file is found, there is no more searching for any further config files.
+1. There is no automatic merging of config files using either `extends` or `overrides`.
+1. When `-c` is passed on the command line, there is no search performed.
+
+### File Pattern Resolution
+
+Because there are file patterns included in `eslint.config.js`, this requires a change to how ESLint decides which files to lint. The process for determining which files to lint is:
+
+1. When a filename is passed directly (such as `eslint foo.js`):
+    1. ESLint checks to see if there is one or more configs where the `files` pattern matches the file that was passed in and does not match the `ignores` pattern. The pattern is evaluated by prepending the directory in which `eslint.config.js` was found to each pattern in `files`. All configs that match `files` and not `ignores` are merged (with the last matching config taking precedence over others). If no config is found, then the file is ignored with an appropriate warning.
+    1. If a matching config is found, then the `ignores` pattern is tested against the filename. If it's a match, then the file is ignored. Otherwise, the file is linted.
+1. When a glob pattern is passed directly (such as `eslint src/*.js`):
+    1. ESLint expands the glob pattern to get a list of files.
+    1. Each file is checked individually as in step 1.
+1. When a directory is passed directly (such as `eslint src`):
```

Comment:
What happens when eslint is executed as `eslint .`? Also, since config now contains globs, should we allow just execution of the eslint without any parameters (so that files to be linted would be resolved from the config globs?).

Reply:
When `eslint .` is run, the `.` is resolved to the current working directory. I can add another step to make that clear.

Before I posted this RFC, my thought was that we could run ESLint without specifying any files or directories, but I removed it because that seemed like it could be confusing. I'm open to adding it back in if it seems useful.

Reply:
If you could. I understand that `.` will just be a glob for current directory, but it's a widely used way of running ESLint, and I think it would be clearer if we mention it explicitly.

In terms of running ESLint without anything, I think it makes sense, given that we can configure globs in the config now.

Reply:
I added a step to clarify what `eslint .` does.

I'm going to add your question about running `eslint` without any files into the Open Questions section to get some more feedback.